packages/template-retail-react-app/app/utils/url.js

```diff
@@ -122,11 +122,12 @@ export const searchUrlBuilder = (searchTerm) => `/search?q=${searchTerm}`
  * Based on your app configuration, this function will replace your current locale shortCode with a new one
  *
  * @param {String} shortCode - The locale short code.
- * @param {Object} [opts] - Options, if there's any.
+ * @param {function(*, *, *, *=): string} - Generates a site URL from the provided href, site and locale.
+ * @param {string[]} opts.disallowParams - URL parameters to remove
  * @param {Object} opts.location - location object to replace the default `window.location`
  * @returns {String} url - The relative URL for the specific locale.
  */
-export const getPathWithLocale = (shortCode, opts = {}) => {
+export const getPathWithLocale = (shortCode, fillUrlTemplate, opts = {}) => {
```

Comment:
This function is used, for the most part, to create urls with a given local, of the locale picker. Could this functionality be done directly using the `fillUrlTemplate` function? Allowing us to delete this entire function? Because it seems complex.

Reply:
We want to find the balance between readability and optimization. 
The `fillUrlTemplate` function has been intentionally simplified/optimized because it's the function we run over and over to generate URLs across the App.  Adding more complex logic will damage the server-side performance optimization.

The `fillUrlTemplate` function expects the first parameter `path` without the `site` or the `locale`.

The function `getPathWithLocale` does extra logic needed only in case we switch locales, before generating the new URL:
1. Remove the `site` and `locale` from the current URL `path`. 
2. Remove the `refine` query string.
3. Create a new URL using `fillUrlTemplate`.

Because switching locales is an edge case of the App URL generation and not the norm. It's best to execute that extra logic only in the specific scenario of switching locales and avoid that logic on the rest of the URLs generated in the App.